### PR TITLE
Allowing custom terraform to refer to remote modules/registries.

### PR DESCRIPTION
**Pull request by juandiegopalomino**
# Description
See the markdown file

# Safety checklist
* [x] This change is backwards compatible and safe to apply by existing users
* [x] This change will NOT lead to data loss
* [x] This change will NOT lead to downtime who already has an env/service setup

## How has this change been tested, beside unit tests?
n/a

### CODE REVIEW

File: modules/custom_terraform/custom_terraform.py [hunk 1]
```diff
@@ -14,12 +15,26 @@ def __init__(self, module: "Module", layer: "Layer"):
...
+            self.module.module_dir_path = os.path.join(path_to_layer, path_to_module)
+        elif path_to_module.startswith("../"):
+            self.module.module_dir_path = os.path.join(path_to_layer, path_to_module)
+        else:
+            # If this is the case, then this refers to a remote module as per
+            # https://www.terraform.io/language/modules/sources
+            self.module.module_dir_path = path_to_module
+            if module_version == "":
+                raise UserErrors(
+                    "module_version must be given to custom terraform for non-local modules"
```
Comment (RemyDeWolf) [inline modules/custom_terraform/custom_terraform.py:32]: do we need to handle this, would terraform handle this logic?
for example, a github repo would not beed a module version but we would throw this error.

File: modules/custom_terraform/custom-terraform.md [hunk 2]
```diff
@@ -135,6 +136,32 @@ resource "google_compute_instance" "default" {
...
+
+```yaml
+environments:
+  - name: aws-example
+    path: "../aws-env.yaml"
+name: baloney
+modules:
+  - type: custom-terraform
+    name: iampolicy
+    path_to_module: "terraform-aws-modules/s3-bucket/aws" # See https://registry.terraform.io/modules/terraform-aws-modules/s3-bucket/aws/latest
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply modules/custom_terraform/custom-terraform.md:152]: yes, that should work too, I just didn't want to do excessive examples

File: modules/custom_terraform/custom_terraform.py (hunk 1 above)
Comment (juandiegopalomino) [reply modules/custom_terraform/custom_terraform.py:32]: I didn't know that, but this is a really nice error message

File: modules/custom_terraform/custom-terraform.yaml [hunk 3]
```diff
@@ -7,6 +7,11 @@ inputs:
     validator: str(required=True)
     description: The path to your terraform module
     default: null
+  - name: version
+    user_facing: true
+    validator: str(required=False)
+    description: The version of the remote module to use
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply modules/custom_terraform/custom-terraform.yaml]: done

File: modules/custom_terraform/custom-terraform.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply modules/custom_terraform/custom-terraform.md]: done

[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply modules/custom_terraform/custom-terraform.md]: done

File: modules/custom_terraform/custom-terraform.md [hunk 4]
```diff
@@ -135,6 +136,32 @@ resource "google_compute_instance" "default" {
...
+    version: "2.13.0" # version needs to be specified for remote modules
+    terraform_inputs:
+      bucket: "dummy-bucket-{aws.account_id}"
+      acl: "private"
+      versioning: 
+        enabled: true
+```
+
+**WARNING** Be very, very, careful about what remote modules you are using, as they leave you wide open to supply chain
+attacks, depending on the security and character of the owner of said module. It's highly advised to use either official
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply modules/custom_terraform/custom-terraform.md]: done

File: modules/custom_terraform/custom_terraform.py (hunk 1 above)
Comment (RemyDeWolf) [reply modules/custom_terraform/custom_terraform.py:36]: here is the use case with github I was trying to explain earlier
```yaml
  - type: custom-terraform
    name: bucket
    source: "github.com/hashicorp/example
    # version not specified
```
https://www.terraform.io/language/modules/sources#github

but this code would throw the error `module_version must be given...`

trying to implement this exception handling seems to be the responsibility of terraform here, right?

File: modules/custom_terraform/gcp-custom-terraform.md [hunk 5]
```diff
@@ -8,8 +8,9 @@ description: Allows user to bring in their own custom terraform module
 ---
 
 This module allows a user to bring in their own custom terraform code into the opta ecosystem, to use in tandem with
-their other opta modules, and even reference them. All a user needs to do is specify the path (absolute or relative to 
-the opta yaml) to your module with the `path_to_module` input, and the desired inputs to your module (if any) via the 
+their other opta modules, and even reference them. All a user needs to do is specify the 
+[path](https://www.terraform.io/language/modules/sources#module-sources)
```
Comment (RemyDeWolf) [inline modules/custom_terraform/gcp-custom-terraform.md:12]: _path_->_source_

File: modules/custom_terraform/gcp-custom-terraform.md [hunk 6]
```diff
@@ -135,20 +136,53 @@ resource "google_compute_instance" "default" {
...
+          first: true
+        }
+      bucket_admins: {
+        second: "user:spam@example.com,eggs@example.com"
+      }
+```
+
+**WARNING** Be very, very, careful about what remote modules you are using, as they leave you wide open to supply chain
+attacks, depending on the security and character of the owner of said module. It's highly advised to use either 
+[official modules](https://registry.terraform.io/browse/modules ) or modules under your company's control.
```
Comment (RemyDeWolf) [inline modules/custom_terraform/gcp-custom-terraform.md:170]: extra space here `/modules )`

File: modules/custom_terraform/aws-custom-terraform.md [hunk 7]
```diff
@@ -138,6 +138,32 @@ When done, you can destroy the custom terraform resource:
...
+    terraform_inputs:
+      bucket: "dummy-bucket-{aws.account_id}"
+      acl: "private"
+      versioning: 
+        enabled: true
+```
+
+**WARNING** Be very, very, careful about what remote modules you are using, as they leave you wide open to supply chain
+attacks, depending on the security and character of the owner of said module. It's highly advised to use either
+[official modules](https://registry.terraform.io/browse/modules ) or modules under your company's control.
```
Comment (RemyDeWolf) [inline modules/custom_terraform/aws-custom-terraform.md:165]: `/modules )` extra space

File: modules/custom_terraform/aws-custom-terraform.md [hunk 8]
```diff
@@ -138,6 +138,32 @@ When done, you can destroy the custom terraform resource:
 opta destroy -c custom-tf.yaml
 ```
 
+## Different options for source
```
Comment (RemyDeWolf) [inline modules/custom_terraform/aws-custom-terraform.md:141]: looking back at it now, may be we could rename the titles to reflect what the sections covers
- Example/Demo -> Use local terraform files
- Different options for source -> Use a remote terraform module
(same comment for GCP)

File: modules/custom_terraform/custom-terraform.yaml [hunk 9]
```diff
@@ -2,11 +2,21 @@ halt: false
 environment_module: true
 is_unique: false
 inputs:
-  - name: path_to_module
+  - name: source
     user_facing: true
-    validator: str(required=True)
+    validator: str(required=False)
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply modules/custom_terraform/custom-terraform.yaml:7]: Neither, see the module processor code. They need to specify at least one

File: modules/custom_terraform/aws-custom-terraform.md (hunk 8 above)
Comment (juandiegopalomino) [reply modules/custom_terraform/aws-custom-terraform.md]: done

File: modules/custom_terraform/aws-custom-terraform.md (hunk 7 above)
Comment (juandiegopalomino) [reply modules/custom_terraform/aws-custom-terraform.md]: done

File: modules/custom_terraform/gcp-custom-terraform.md (hunk 6 above)
Comment (juandiegopalomino) [reply modules/custom_terraform/gcp-custom-terraform.md]: done

File: modules/custom_terraform/gcp-custom-terraform.md (hunk 5 above)
Comment (juandiegopalomino) [reply modules/custom_terraform/gcp-custom-terraform.md]: done

File: modules/custom_terraform/custom-terraform.json [hunk 10]
```diff
@@ -3,15 +3,23 @@
   "description": "To bring in your own custom terraform module",
   "type": "object",
   "properties": {
-    "path_to_module": {
+    "source": {
       "type": "string",
       "description": "The path to your terraform module"
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply modules/custom_terraform/custom-terraform.json:8]: done

File: modules/custom_terraform/custom-terraform.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (RemyDeWolf) [reply modules/custom_terraform/custom-terraform.yaml:7]: ok, is it for backward compatibility?

**Review by RemyDeWolf — APPROVED** [review]
looks great

**Pull request merged**